lib/render/ParticleRenderer.cpp

```diff
@@ -0,0 +1,179 @@
+    float         dirScale = p->GetValueDouble(ParticleParams::DirectionScaleTag, 1);
+    vector<Grid *> vecGrids;
+    if (showDir) {
+        vector<string> vecNames = p->GetFieldVariableNames();
+        vector<string> mainVarCoords;
+        _dataMgr->GetVarCoordVars(varName, true, mainVarCoords);
+
+        for (auto var : vecNames) {
+            vector<string> varCoords;
+            _dataMgr->GetVarCoordVars(var, true, varCoords);
+
+            if (mainVarCoords != varCoords) {
+                if (grid) {
+                    _dataMgr->UnlockGrid(grid);
+                    delete grid;
+                }
+                for (auto g : vecGrids) {
+                    _dataMgr->UnlockGrid(g);
+                    delete g;
+                }
+                SetErrMsg("Variable \"%s\" on different grid from main variable", var.c_str());
+                return -1;
+            }
+
+            Grid *ng = _dataMgr->GetVariable(p->GetCurrentTimestep(), var, p->GetRefinementLevel(), p->GetCompressionLevel(), minExt, maxExt, true);
+            if (!ng) {
+                if (grid) {
+                    _dataMgr->UnlockGrid(grid);
+                    delete grid;
+                }
+                for (auto g : vecGrids) {
+                    _dataMgr->UnlockGrid(g);
+                    delete g;
+                }
+                SetErrMsg("Cannot read var \"%s\"", var.c_str());
+                return -1;
+            }
+
+            vecGrids.push_back(ng);
+        }
+    }
+
+    vector<size_t> dims = grid->GetDimensions();
+    size_t         nCoords = 1;
+    for (const auto d : dims) nCoords *= d;
+
+    auto *lgl = _glManager->legacy;
+    lgl->Color3f(1, 1, 1);
+    lgl->Begin(showDir ? GL_LINES : GL_POINTS);
+
+    long                        renderedParticles = 0;
+    auto                        node = grid->ConstNodeBegin(minExt, maxExt);
+    auto                        nodeEnd = grid->ConstNodeEnd();
+    vector<double>              coordsBuf(3);
+    vector<Grid::ConstIterator> dirs;
+    for (auto g : vecGrids) dirs.push_back(g->cbegin(minExt, maxExt));
+    for (size_t i = 0; node != nodeEnd; ++node, ++i) {
+        if (i % stride) {
+            if (showDir)
+                for (auto &it : dirs) ++it;
+            continue;
+        }
+
+        const float value = grid->GetValueAtIndex(*node);
+        grid->GetUserCoordinates(*node, coordsBuf);
+        const glm::vec3 p(coordsBuf[0], coordsBuf[1], coordsBuf[2]);
+
+        lgl->Color4fv(&LUT[4 * glm::clamp((int)(255 * (value - mapMin) / mapDif), 0, 255)]);
+        lgl->Vertex3fv((float *)&p);
+        renderedParticles++;
+
+        if (showDir) {
+            const glm::vec3 n(*(dirs[0]), *(dirs[1]), *(dirs[2]));
+            const glm::vec3 p2 = p + n * dirScale;
+            lgl->Vertex3fv((float *)&p2);
+
+            for (auto &it : dirs) ++it;
+        }
+    }
+    lgl->End();
+
+    //    printf("Rendered %li particles\n", renderedParticles);
```

Comment:
do you want to delete the debug code?

Reply:
I would like to leave this for later work.